consumer/transaction.go

```diff
@@ -187,80 +201,98 @@ func txnRead(s *shard, txn, prev *transaction, env EnvelopeOrError, ok bool) err
 		txn.readCh = nil // Channel is closed, don't select it again.
 		return nil
 	} else if env.Error != nil {
-		return errors.WithMessage(env.Error, "readMessage")
+		return fmt.Errorf("readMessage: %w", env.Error)
 	}
-	txn.readThrough[env.Journal.Name] = env.End
 
 	// DEPRECATED metrics to be removed:
 	bytesConsumedTotal.Add(float64(env.End - env.Begin))
 	readHeadGauge.WithLabelValues(env.Journal.Name.String()).Set(float64(env.End))
 	// End DEPRECATED metrics.
 
-	if !s.sequencer.QueueUncommitted(env.Envelope) {
-		// Envelope is a continuation of an ongoing sequence of messages.
-		// No need to dequeue for committed messages (there are none).
-		return nil
-	}
+	// |env| is read-uncommitted. Queue and act on its sequencing outcome.
+	var outcome = s.sequencer.QueueUncommitted(env.Envelope)
 
-	for {
-		switch deq, err := s.sequencer.DequeCommitted(); err {
-		case nil:
-			if err = txnConsume(s, txn, deq); err != nil {
-				return err
-			}
+	switch outcome {
```

Comment:
There's a lot of these `QueueOutcome` constants, and it's kind of tough as a reader to see that we're handling all of these correctly in terms of the contract with the `Sequencer`. One suggestion would be to define functions like `func (o QueueOutcome) MustStep() bool` make that more clear.

Reply:
You know, I had done that but ended up backing it out, because you're always allowed to call `Step` and the simpler cases just do that. Otherwise the only QueueOutcome you need to pay attention to is `QueueAckCommitReplay`, because its on you to start a replay read. `transaction.go` cares about more of these, because it's doing fancier stuff with respect to reporting progress. So I backed it out because there isn't actually a common thread on how it'd be used.